libcudacxx/include/cuda/std/__limits/numeric_limits_ext.h

```diff
@@ -644,6 +644,105 @@ class __numeric_limits_impl<__nv_fp4_e2m1, __numeric_limits_type::__floating_poi
 };
 #endif // _CCCL_HAS_NVFP4_E2M1()
 
+// __fp128
+
+#if _CCCL_HAS_FLOAT128()
+template <>
+class __numeric_limits_impl<__float128, __numeric_limits_type::__floating_point>
+{
+public:
+  using type = __float128;
+
+  static constexpr bool is_specialized = true;
+
+  static constexpr bool is_signed   = true;
+  static constexpr int digits       = 113;
+  static constexpr int digits10     = 33;
+  static constexpr int max_digits10 = 36;
+  _LIBCUDACXX_HIDE_FROM_ABI static constexpr type min() noexcept
+  {
+    return 3.36210314311209350626267781732175260e-4932q;
+  }
+  _LIBCUDACXX_HIDE_FROM_ABI static constexpr type max() noexcept
+  {
+    return 1.18973149535723176508575932662800702e+4932q;
+  }
+  _LIBCUDACXX_HIDE_FROM_ABI static constexpr type lowest() noexcept
+  {
+    return -max();
+  }
+
+  static constexpr bool is_integer = false;
+  static constexpr bool is_exact   = false;
+  static constexpr int radix       = __FLT_RADIX__;
+  _LIBCUDACXX_HIDE_FROM_ABI static constexpr type epsilon() noexcept
+  {
+    return 1.92592994438723585305597794258492732e-34q;
+  }
+  _LIBCUDACXX_HIDE_FROM_ABI static constexpr type round_error() noexcept
+  {
+    return 0.5q;
+  }
+
+  static constexpr int min_exponent   = -16381;
+  static constexpr int min_exponent10 = -4931;
+  static constexpr int max_exponent   = 16384;
+  static constexpr int max_exponent10 = 4932;
+
+  static constexpr bool has_infinity             = true;
+  static constexpr bool has_quiet_NaN            = true;
+  static constexpr bool has_signaling_NaN        = true;
+  static constexpr float_denorm_style has_denorm = denorm_present;
+  static constexpr bool has_denorm_loss          = false;
+
+#  if defined(_CCCL_BUILTIN_HUGE_VALF128)
+  _LIBCUDACXX_HIDE_FROM_ABI static constexpr type infinity() noexcept
+  {
+    return _CCCL_BUILTIN_HUGE_VALF128();
+  }
+#  else // ^^^ _CCCL_BUILTIN_HUGE_VALF128 ^^^ // vvv !_CCCL_BUILTIN_HUGE_VALF128 vvv
+  _LIBCUDACXX_HIDE_FROM_ABI static _LIBCUDACXX_CONSTEXPR_BIT_CAST type infinity() noexcept
+  {
+    return _CUDA_VSTD::bit_cast<type>(__uint128_t{0x7fff000000000000} << 64);
```

Comment:
Maybe we can improve the readability a bit by using digit separators:
```suggestion
    return _CUDA_VSTD::bit_cast<type>(__uint128_t{0x7fff'0000'0000'0000} << 64);
```
or shifting by a larger amount:
```suggestion
    return _CUDA_VSTD::bit_cast<type>(__uint128_t{0x7fff} << 64+48);
```